llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -1180,6 +1180,82 @@ Operation *ReverseOp::getTiledImplementation(OpBuilder &builder,
   return tiledRevOp;
 }
 
+//===----------------------------------------------------------------------===//
+// TopkOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult TopkOp::verify() {
+  Operation *op = getOperation();
+  if (getNumInputs() != 2) {
+    return op->emitOpError("expected two input operands");
+  }
+  if (getNumOutputs() != 2) {
+    return op->emitOpError("expected two output operands");
+  }
+  if (!inputValues().getType().isa<ShapedType>() ||
+      !inputIndices().getType().isa<ShapedType>()) {
+    return op->emitOpError("expected input element types to be shaped");
+  }
+  if (!outputValues().getType().isa<ShapedType>() ||
+      !outputIndices().getType().isa<ShapedType>()) {
+    return op->emitOpError("expected output element types to be shaped");
+  }
+  // Ensure input/output values types match
+  ShapedType inputValuesType = inputValues().getType().cast<ShapedType>();
+  ShapedType outputValuesType = outputValues().getType().cast<ShapedType>();
+  if (inputValuesType.getElementType() != outputValuesType.getElementType()) {
+    return op->emitOpError("expected input/output value types to be identical");
+  }
+  // Indices must be index or float
+  ShapedType inputIndicesType = inputIndices().getType().cast<ShapedType>();
+  ShapedType outputIndicesType = outputIndices().getType().cast<ShapedType>();
+  if (!inputIndicesType.getElementType().isIntOrIndex() ||
+      !outputIndicesType.getElementType().isIntOrIndex()) {
+    return op->emitOpError(
+        "expected input/output indices types to be int or index");
+  }
+  // Ranks must match
+  if (inputValuesType.getRank() != outputValuesType.getRank() ||
+      inputIndicesType.getRank() != outputIndicesType.getRank()) {
+    return op->emitOpError("expected input/output to have the same rank");
+  }
+  // Input indicies/values and output indices/values should have the same shape
+  if (llvm::any_of(
+          llvm::zip(inputValuesType.getShape(), inputIndicesType.getShape()),
+          [](std::tuple<int64_t, int64_t> s) {
+            return std::get<0>(s) != ShapedType::kDynamicSize &&
+                   std::get<1>(s) != ShapedType::kDynamicSize &&
+                   std::get<0>(s) != std::get<1>(s);
+          })) {
+    return op->emitOpError("input indicies/values shape must match");
+  }
+  if (llvm::any_of(
+          llvm::zip(outputValuesType.getShape(), outputIndicesType.getShape()),
+          [](std::tuple<int64_t, int64_t> s) {
+            return std::get<0>(s) != ShapedType::kDynamicSize &&
+                   std::get<1>(s) != ShapedType::kDynamicSize &&
+                   std::get<0>(s) != std::get<1>(s);
```

Comment:
It shows up three times in the check. We can create a lambda or static function for this check.